### ISSUE TITLE
Support for rollout restart of deployment/daemonset/statefulset when secret changes

### ISSUE THREAD

**Issue by gmauleon** [external-secrets/external-secrets#3218]
Related to this closed issue https://github.com/external-secrets/external-secrets/issues/1865

Just taking a chance, we were looking at existing secrets controllers out there and specifically the `vault secret operator` and of course `external secrets` which seems to be the top contenders.

External secrets seems to tick all the boxes and is more versatile, but one very useful feature of the former is the ability to trigger rollout after secret changes (spec [here](https://developer.hashicorp.com/vault/docs/platform/k8s/vso/api-reference#rolloutrestarttarget) )

I know, I know, we could install and managed yet another controller to do that, but since the Vault secret operator is fairly new and now supports it, I was wondering if this is a feature that should be reconsidered for `external-secrets`? 

As I said taking a chance 😁

Thanks in advance for your answers!

**Comment by Cib0rg** [comment]
It works like a charm with stakater reloader, just add metadata to secrets template.

**Comment by ChristianCiach** [comment]
It's true that the stakater reloader works like a charm. It's also true that this missing functionality of external-secrets was the sole reason for our decision to chose the Hashicorp VSO, mainly due to easy of deployment. We have many clusters, and any additional operator comes with additional operational cost (maintenance, testing).

**Comment by Cib0rg** [comment]
Agreed. I just tried to help you until (and if) this function will be added.

**Comment by ait-maksymp** [comment]
Unfortunately, stakater reloader thinks that a Secret was _changed_ every time it is _updated_ (polled from AWS Secrets without actually changing the contents) by the external-secrets, and I get a rolling restart every `refreshInterval`.

**Comment by asger-noer** [comment]
I'd love to have this feature available in the ESO. It seems like the industry is leaning towards it being the secrets controllers's responsibility to rollout a restart if the underlying data have changed. I don't know if there would be a clever way of only doing a restart of a deployment/statefulset if the `spec.template.spec.containers.envFrom.secretRef` reference is managed by ESO.

**Comment by thesuperzapper** [comment]
> Unfortunately, stakater reloader thinks that a Secret was _changed_ every time it is _updated_ (polled from AWS Secrets without actually changing the contents) by the external-secrets, and I get a rolling restart every `refreshInterval`.

@ait-maksymp it's likely that once https://github.com/external-secrets/external-secrets/pull/4086 is merged, your Kyverno/other systems won't constantly trigger restarts for secrets managed by ExternalSecret resources.

That is, unless the secret is actually changing, e.g. if you have two ExternalSecrets managing the same Secret, or are using a [generator](https://external-secrets.io/v0.10.5/guides/generator/) with `refreshInterval > 0`.

**Comment by ait-maksymp** [comment]
> That is, unless the secret is actually changing, e.g. if you have two ExternalSecrets managing the same Secret

Oh, this is exactly what I am doing, I merge contents of two ExternalSecrets into one Secret 🤔
Thank you @thesuperzapper for food for thought.

**Comment by danielloader** [comment]
> It works like a charm with stakater reloader, just add metadata to secrets template.

Can you explain what you need to add to the template to make reloader work?

**Comment by gusfcarvalho** [comment]
> > It works like a charm with stakater reloader, just add metadata to secrets template.
> 
> Can you explain what you need to add to the template to make reloader work?

that's documented in their project 😄 . Annotations need to be added to any deployments using the Secret you want to trigger auto rollout

**Comment by evrardj-roche** [comment]
I think we can close this... ESO's project current answer is "use stakater/reloader".
